### ISSUE TITLE
Memoization yields unexpected results

### ISSUE THREAD

**Issue by kruvcraft21** [vfsfitvnm/frida-il2cpp-bridge#90]
Hi,
after upgrading to 0.7.1 when using this code

``` ts
Il2Cpp.Domain.assembly("Assembly").image.class("TestClass").method<void>("Method").implementation = function(p1: Il2Cpp.Object) : void{
    this.method<void>("Method").invoke(p1)
    this.method<void>("Method").invoke(p1)
}
```

I sometimes get this error

`Error: already replaced this function
    at Object.value [as replace] (frida/runtime/core.js:373:1)
    at set implementation [as implementation] (node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:87:1)`

previous versions didn't have this. At the same time, the method does not work as I wrote

**Comment by kruvcraft21** [comment]
I'm sorry I didn't figure out the error, now I have new information

I have two classes A and B

```csharp
class A: System.Object {
    System.Void M(C c); //0x01d56aac
}

class B: A {
    System.Void M(C c); // 0x01d56e60
}
```

if you write this code

```typescript
Il2Cpp.Domain.assembly("Assembly").image.class("A").method<void>("M").implementation = function(c: Il2Cpp.Object) : void{
    this.method<void>("M").invoke(c)
}

Il2Cpp.Domain.assembly("Assembly").image.class("B").method<void>("M").implementation = function(c: Il2Cpp.Object) : void{
    this.method<void>("M").invoke(c)
}
```

Only after this will an error appear

`Error: already replaced this function
    at Object.value [as replace] (frida/runtime/core.js:373:1)
    at set implementation [as implementation] (node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:87:1)`

**Comment by vfsfitvnm** [comment]
Hmm this is strange; would you post the dump of these two classes (just console.log them)?

**Comment by vfsfitvnm** [comment]
Oh, I could reproduce using this snippet:
```ts
const SystemString = Il2Cpp.Image.corlib.class("System.String");
const SystemObject = SystemString.parent!;

console.log(SystemObject.method("GetHashCode"));
console.log(SystemString.method("GetHashCode"));

console.log(SystemString.methods.find(e => e.name == "GetHashCode"));
```
Which outputs:
```cs
System.Int32 GetHashCode(); // 0x00c7d330 <--- correct
System.Int32 GetHashCode(); // 0x00c7d330 <--- wrong
System.Int32 GetHashCode(); // 0x00d0aff0 <--- correct
```
However, if the exact number of arguments is passed (the default is -1), then it works fine:
```ts
console.log(SystemString.method("GetHashCode", 0));
```
```cs
System.Int32 GetHashCode(); // 0x00c7d330 <--- correct
```

I didn't realize this...

**Comment by vfsfitvnm** [comment]
This is awkward, as this is the relevant source code:
```cpp
const int Class::IgnoreNumberOfArguments = -1;

const MethodInfo* Class::GetMethodFromName(Il2CppClass *klass, const char* name, int argsCount)
{
    return GetMethodFromNameFlags(klass, name, argsCount, 0);
}

const MethodInfo* Class::GetMethodFromNameFlags(Il2CppClass *klass, const char* name, int argsCount, int32_t flags)
{
    Class::Init(klass);

    while (klass != NULL)
    {
        void* iter = NULL;
        while (const MethodInfo* method = Class::GetMethods(klass, &iter))
        {
            if (method->name[0] == name[0] &&
                !strcmp(name, method->name) &&
                (argsCount == IgnoreNumberOfArguments || method->parameters_count == argsCount) &&
                ((method->flags & flags) == flags))
            {
                return method;
            }
        }

        klass = klass->parent;
    }

    return NULL;
}
```

**Comment by vfsfitvnm** [comment]
Ok lol, there is a problem on my side... this one works fine:
```ts
console.log(new Il2Cpp.Method(Il2Cpp.Api._classGetMethodFromName(SystemString, Memory.allocUtf8String("GetHashCode"), -1))); 
```
Look at this:
```ts
console.log("System.String ->", SystemString.method("ToCharArray"));
console.log("System.Object ->", SystemObject.method("ToCharArray")); 
```
This logs:
```cs
System.Char[] ToCharArray(); // 0x00d0aec0 <--- correct
System.Char[] ToCharArray(); // 0x00d0aec0 <--- wtf? there's no such method in System.Object
```

The problem is definitely here: https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/utils/utils.ts#L35

**Comment by kruvcraft21** [comment]
I'm sorry I didn't answer, as I understood the problem is that I'm trying to change a method that doesn't belong to a child class. Do you need a dump of these classes ?

**Comment by vfsfitvnm** [comment]
The problem is in the "cache system" I use to avoid calling the same native function with the same arguments over and over. I just call it once and then save the result in a map. This is called memoization.
https://www.better.dev/understanding-memoization-in-javascript

**Comment by vfsfitvnm** [comment]
Ok, I found a good package for this job; however, after a further investigation, it looks like this isn't required at all (a proper memoization function is even less performant), so I guess I'm removing it

**Comment by vfsfitvnm** [comment]
Fixed in 0.7.2
